pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller.go

```diff
@@ -0,0 +1,309 @@
+type APIExportEndpointSliceStatus = apisv1alpha1.APIExportEndpointSliceStatus
+type Patcher = apisv1alpha1client.APIExportEndpointSliceInterface
+type Resource = committer.Resource[*APIExportEndpointSliceSpec, *APIExportEndpointSliceStatus]
+type CommitFunc = func(context.Context, *Resource, *Resource) error
+
+// controller reconciles APIExportEndpointSlices. It ensures that the shard endpoints are populated
+// in the status of every APIExportEndpointSlices.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	listShards                            func() ([]*corev1alpha1.Shard, error)
+	listAPIExportEndpointSlices           func() ([]*apisv1alpha1.APIExportEndpointSlice, error)
+	getAPIExportEndpointSlice             func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExportEndpointSlice, error)
+	getAPIExport                          func(path logicalcluster.Path, name string) (*apisv1alpha1.APIExport, error)
+	getAPIExportEndpointSlicesByAPIExport func(key string) ([]interface{}, error)
+
+	commit CommitFunc
+}
+
+// enqueueAPIExportEndpointSlice enqueues an APIExportEndpointSlice.
+func (c *controller) enqueueAPIExportEndpointSlice(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).Info("queueing APIExportEndpointSlice")
+	c.queue.Add(key)
+}
+
+// enqueueAPIExportEndpointSlicesForAPIExport enqueues APIExportEndpointSlices referencing a specific APIExport.
+func (c *controller) enqueueAPIExportEndpointSlicesForAPIExport(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	objectsForAPIExport, err := c.getAPIExportEndpointSlicesByAPIExport(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), obj.(*apisv1alpha1.APIExport))
+	logging.WithQueueKey(logger, key).V(2).Info("queuing APIExportEndpointSlices because of referenced APIExport")
+	for _, obj := range objectsForAPIExport {
+		slice, ok := obj.(*apisv1alpha1.APIExportEndpointSlice)
+		if !ok {
+			runtime.HandleError(fmt.Errorf("obj is supposed to be an APIExportEndpointSlice, but is %T", obj))
+			return
+		}
+		c.enqueueAPIExportEndpointSlice(slice)
+	}
+}
+
+// enqueueAllAPIExportEndpointSlices enqueues all APIExportEndpointSlices.
+func (c *controller) enqueueAllAPIExportEndpointSlices(shard interface{}) {
+	list, err := c.listAPIExportEndpointSlices()
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), shard.(*corev1alpha1.Shard))
+	for i := range list {
+		key, err := kcpcache.MetaClusterNamespaceKeyFunc(list[i])
```

Comment:
not `DeletionHandlingMetaClusterNamespaceKeyFunc` ?

Reply:
thanks. done.